Add `EntityRefExcept` and `EntityMutExcept` world queries, in preparation for generalized animation.

pcwalton opened the pull request:
This commit adds two new `WorldQuery` types: `EntityRefExcept` and `EntityMutExcept`. These types work just like `EntityRef` and `EntityMut`, but they prevent access to a statically-specified list of components. For example, `EntityMutExcept<(AnimationPlayer, Handle<AnimationGraph>)>` provides mutable access to all components except for `AnimationPlayer` and `Handle<AnimationGraph>`. These types are useful when you need to be able to process arbitrary queries while iterating over the results of another `EntityMut` query.

The motivating use case is *generalized animation*, which is an upcoming feature that allows animation of any component property, not just rotation, translation, scaling, or morph weights. To implement this, we must change the current `AnyOf<(&mut Transform, &mut MorphWeights)>` to instead be `EntityMutExcept<(AnimationPlayer, Handle<AnimationGraph>)>`. It's possible to use `FilteredEntityMut` in conjunction with a dynamically-generated system instead, but `FilteredEntityMut` isn't optimized for the use case of a large number of allowed components coupled with a small set of disallowed components. No amount of optimization of `FilteredEntityMut` produced acceptable performance on the `many_foxes` benchmark. `Query<EntityMut, Without<AnimationPlayer>>` will not suffice either, as it's legal and idiomatic for an `AnimationTarget` and an `AnimationPlayer` to coexist on the same entity.

An alternate proposal was to implement a somewhat-more-general `Except<Q, CL>` feature, where Q is a `WorldQuery` and CL is a `ComponentList`. I wasn't able to implement that proposal in a reasonable way, because of the fact that methods like `EntityMut::get_mut` and `EntityRef::get` are inherent methods instead of methods on `WorldQuery`, and therefore there was no way to delegate methods like `get` and `get_mut` to the inner query in a generic way. Refactoring those methods into a trait would probably be possible. However, I didn't see a use case for a hypothetical `Except` with arbitrary queries: `Query<Except<(&Transform, &Visibility), Visibility>>` would just be a complicated equivalent to `Query<&Transform>`, for instance. So, out of a desire for simplicity, I omitted a generic `Except` mechanism.

I've tested the performance of generalized animation on `many_foxes` and found that, with this patch, `animate_targets` has a 7.4% slowdown over `main`. With `FilteredEntityMut` optimized to use `Arc<Access>`, the slowdown is 75.6%, due to contention on the reference count. Without `Arc<Access>`, the slowdown is even worse, over 2x.

## Testing

New tests have been added that check that `EntityRefExcept` and `EntityMutExcept` allow and disallow access to components properly and that the query engine can correctly reject conflicting queries involving those types.

A Tracy profile of `many_foxes` with 10,000 foxes showing generalized animation using `FilteredEntityMut` (red) vs. main (yellow) is as follows:

![Screenshot 2024-09-12 225914](https://github.com/user-attachments/assets/2993d74c-a513-4ba4-85bd-225672e7170a)

A Tracy profile of `many_foxes` with 10,000 foxes showing generalized animation using this `EntityMutExcept` (yellow) vs. main (red) is as follows:

![Screenshot 2024-09-14 205831](https://github.com/user-attachments/assets/4241015e-0c5d-44ef-835b-43f78a24e604)

hymm commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -2998,6 +3254,187 @@ mod tests {
+
+        world.run_system_once(system);
+    }
+
```
We should have tests that all the get_* methods return none when the component exists and is in the deny list, since this is a safety concern.

pcwalton replied:
Added.

SkiFire13 commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -810,11 +811,215 @@ unsafe impl<'a> WorldQuery for FilteredEntityMut<'a> {
+        state: &Self::State,
+        filtered_access: &mut FilteredAccess<ComponentId>,
+    ) {
+        // `Access` specifies which components are allowed, not which are
+        // disallowed. So we iterate through all component IDs in the `World`,
+        // except for the excluded ones, and add them one-by-one to the
+        // `Access`. This is somewhat slow, but this code shouldn't be a hot
+        // path.
+        for component_index in 0..state.component_count {
+            let component_id = ComponentId::new(component_index);
+            if !state.ids.contains(component_id) {
+                filtered_access
+                    .access_mut()
+                    .add_component_read(component_id);
+            }
+        }
```
This seems pretty suspicious. What happens when a component is initialized after `update_component_access` is called for this query?

SkiFire13 replied:
Yeah this seems to be unsound

```rs
    #[test]
    fn bad() {
        let mut world = World::new();
        world.init_component::<TestComponent>();

        let system = (
            QueryParamBuilder::new(|_| {}),
            QueryParamBuilder::new(|_| {}),
        )
            .build_state(&mut world)
            .build_system(
                |mut q1: Query<EntityMutExcept<TestComponent>>,
                 mut q2: Query<&mut TestComponent2>| {
                    let mut e1 = q1.single_mut();
                    let mut c1 = e1.get_mut::<TestComponent2>().unwrap();
                    let mut c2 = q2.single_mut();

                    assert!(!std::ptr::eq(&mut *c1, &mut *c2));
                },
            );

        world.spawn(TestComponent2(0));

        world.run_system_once(system);
    }
```

pcwalton replied:
The maximum component index is no longer relevant.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -2384,6 +2384,259 @@ pub enum TryFromFilteredError {
+    /// Returns `None` if the component doesn't have a component of that type or
+    /// if the type is one of the excluded components.
+    #[inline]
+    pub fn get_mut<C>(&self) -> Option<Mut<'_, C>>
```
```suggestion
    pub fn get_mut<C>(&mut self) -> Option<Mut<'_, C>>
```

pcwalton replied:
Fixed.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -2384,6 +2384,259 @@ pub enum TryFromFilteredError {
+
+            #[inline]
+            fn contains<C>() -> bool where C: Component {
+                true $(|| TypeId::of::<C>() == TypeId::of::<$name>())*
```
This will always return `true`, I guess you meant to use `false` here?

pcwalton replied:
I removed `ComponentList`.

alice-i-cecile reviewed (commented):
I like the concept and naming here: `Access` is heavy, and the perf results are compelling.

>  We could perhaps change get_component_ids to get_component_type_ids and implement get_component_ids on top of it

I liked your suggestion here on [Discord](https://discordapp.com/channels/691052431525675048/749335865876021248/1284798340915265608). I think that as long as `get_component_type_ids` returns a partial list, which we then augment with the untyped components, we should be fine.

Doing this should simplify the implementation a fair bit, but let me know if you'd prefer to leave this for follow-up.

chescock commented:
Would it make sense to modify `Access` to be able to express "all components except for a set"?  

That could cleanly solve the issue @SkiFire13 [noted above](https://github.com/bevyengine/bevy/pull/15207#discussion_r1759987340), since `QueryState::update_archetype_component_access()` could evaluate new components against the `Access`.  

That could be done by replacing the `reads_all_components` and `writes_all_components` flags with ones that invert the meaning of the corresponding `FixedBitSet`.  Methods like `has_component_write()` would use `invert_component_writes ^ component_writes.contains(index)` instead of `writes_all || component_writes.contains(index)`.  I believe `FixedBitSet` exposes enough bitwise operations to implement all four cases for methods like `extend()` and `is_compatible()`.

chescock commented:
It might be more general to use `QueryData` instead of `ComponentList`.  Then an `EntityMutExcept<&mut T>` could have no access to `T`, while an `EntityMutExcept<&T>` could still read it.

chescock commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -2384,6 +2384,259 @@ pub enum TryFromFilteredError {
+    }
+}
+
+impl_component_list_tuple!(C1);
```
Bevy has an `all_tuples!` macro that helps with this.  It also makes it easy to add `#[doc(fake_variadic)]`, which you may want to do here.

pcwalton replied:
I removed `ComponentList`.

pcwalton commented:
> It might be more general to use `QueryData` instead of `ComponentList`. Then an `EntityMutExcept<&mut T>` could have no access to `T`, while an `EntityMutExcept<&T>` could still read it.

I think QueryData would require messing with Access in the get method, which would defeat the purpose of EntityRefExcept as Access is too slow. I'd prefer to just replace ComponentList with Bundle.

SkiFire13 commented:
I tried seeing if @chescock's proposal would work, and it seems possible to me though there are some details that need to be refined (e.g. replace all uses of `Access::component_reads_and_writes`, which is no longer possible to implement). You can see my try at https://github.com/bevyengine/bevy/compare/main...SkiFire13:bevy:access-inverted

pcwalton commented:
I've implemented inverted bit sets. All ECS tests pass. I added more tests to make sure `get`/`get_ref`/`get_mut` are correct.

I also removed `ComponentList` in favor of `Bundle`. This *is* about 5% slower in `many_foxes` than `ComponentList` is, but it's probably worth the extra simplicity.

Unfortunately, fixing the `contains` method caused a significant regression:

![Screenshot 2024-09-16 023649](https://github.com/user-attachments/assets/981718ef-4e4d-4951-9bdb-bc10cb1b8e2f)

I don't have any ideas to eliminate the regression at this point, because we have to either check component IDs or type IDs no matter what, so I'm going to argue in favor of just taking the regression.

This should be ready for review again.

chescock commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -630,22 +632,22 @@ unsafe impl<'a> WorldQuery for FilteredEntityRef<'a> {
+        for id in archetype.components() {
+            if state.access.has_component_read(id) {
                 access.add_component_read(id);
             }
-        });
+        }
         fetch.1 = access;
     }
 
     #[inline]
     unsafe fn set_table<'w>(fetch: &mut Self::Fetch<'w>, state: &Self::State, table: &'w Table) {
         let mut access = Access::default();
-        state.access.component_reads().for_each(|id| {
-            if table.has_column(id) {
+        for id in table.columns() {
+            if state.access.has_component_read(id) {
                 access.add_component_read(id);
             }
-        });
+        }
         fetch.1 = access;
```
Yeah, the reason it was originally written like this was fixed in #13554.  

I think it can even be `fetch.1.clone_from(&state.access);` to make allocation here less likely.

pcwalton replied:
I made that change.

chescock commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -508,22 +508,26 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
-                    access.add_component_write(id);
-                }
-            });
+        for component_id in archetype.components() {
+            let Some(archetype_component_id) = archetype.get_archetype_component_id(component_id)
+            else {
+                continue;
+            };
```
We could avoid the lookup by adding a method on `Archetype` like

```rust
pub(crate) fn components_with_archetype_component_id(
    &self,
) -> impl Iterator<Item = (ComponentId, ArchetypeComponentId)> + '_ {
    self.components
        .iter()
        .map(|(component_id, info)| (*component_id, info.archetype_component_id))
}
```

pcwalton replied:
I added that method.

chescock commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -508,22 +508,26 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
-                    access.add_component_write(id);
-                }
-            });
+        for component_id in archetype.components() {
```
I think we expect archetypes to usually have more components than queries, so flipping the iteration direction unconditionally could be a performance regression.  It might be worth doing 

```rust
if self.component_access.access.component_reads_and_writes_inverted() {
    // New code
} else {
    // Old code
}
```

But I haven't profiled it, so it may not matter at all.

chescock reviewed (approved):
We should fix the [unsoundness issue](https://github.com/bevyengine/bevy/pull/15207#discussion_r1761004132) before merging, but I'm happy to make or review a follow-up PR for the rest of the cleanup that @SkiFire13 recommends if @pcwalton wants to be done shaving this particular yak.

chescock commented on crates/bevy_ecs/src/query/access.rs:
```diff
@@ -577,6 +734,33 @@ impl<T: SparseSetIndex> Access<T> {
+    /// reads and writes or cannot read or write, depending on the value of
+    /// [`Access::component_reads_and_writes_inverted`].
+    ///
+    /// Because this method depends on internal implementation details of
```
I think this method could be made safe to use if it returned an `enum` with variants representing "these components" and "everything except for these components".

pcwalton commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -815,6 +813,186 @@ unsafe impl<'a> QueryData for FilteredEntityMut<'a> {
+        let access = filtered_access.access_mut();
+        access.write_all_components();
+        for id in state {
+            access.remove_component_read(*id);
```
Are you sure? I wrote `remove_component_read` there in order to remove both reads and writes.

SkiFire13 replied:
Oh right, I was confused by the swapped behaviour compared to adding read/writes.

pcwalton commented on crates/bevy_ecs/src/query/access.rs:
```diff
@@ -577,6 +734,33 @@ impl<T: SparseSetIndex> Access<T> {
+    #[doc(hidden)]
+    pub fn component_reads_and_writes_inverted(&self) -> bool {
+        self.component_read_and_writes_inverted
+    }
```
Done. I did have to add `#[allow(deprecated)]` in a couple of places to allow us to call it.

pcwalton commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -815,6 +813,186 @@ unsafe impl<'a> QueryData for FilteredEntityMut<'a> {
+        access.read_all_components();
+        for id in state {
+            access.remove_component_read(*id);
+        }
```
Fixed both issues, and added a test for the latter.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -3188,11 +3188,35 @@ mod tests {
+        // This should panic, because we have a mutable borrow on
+        // `TestComponent` but have a simultaneous indirect immutable borrow on
+        // that component via `EntityRefExcept`.
+        let system = (QueryParamBuilder::new(|builder| {
+            builder.data::<(&mut TestComponent, EntityRefExcept<TestComponent2>)>();
+        }),)
+            .build_state(&mut world)
+            .build_system(|_: Query<(&mut TestComponent, EntityRefExcept<TestComponent2>)>| {});
```
FYI this is no different than a plain system taking a `Query<(&mut TestComponent, EntityRefExcept<TestComponent2>)>`. `builder.data` is doing nothing here because the data for the static type in the query is pre-registered.

alice-i-cecile replied:
Yeah, world.run_system_once is probably going to be easier.

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -3278,11 +3302,35 @@ mod tests {
+        // This should panic, because we have a mutable borrow on
+        // `TestComponent` but have a simultaneous indirect immutable borrow on
+        // that component via `EntityRefExcept`.
+        let system = (QueryParamBuilder::new(|builder| {
+            builder.data::<(&mut TestComponent, EntityMutExcept<TestComponent2>)>();
+        }),)
+            .build_state(&mut world)
+            .build_system(|_: Query<(&mut TestComponent, EntityMutExcept<TestComponent2>)>| {});
```
Same here

SkiFire13 commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -3188,11 +3188,35 @@ mod tests {
+    #[should_panic]
+    fn entity_ref_except_conflicts_with_self() {
+        let mut world = World::new();
+        world.init_component::<TestComponent>();
+        world.init_component::<TestComponent2>();
```
There's no need to explicitly initialize the components.

SkiFire13 commented on crates/bevy_ecs/src/query/access.rs:
```diff
@@ -185,6 +205,35 @@ impl<T: SparseSetIndex> Access<T> {
+        }
+    }
+
+    /// Removes both read and write access to the component given by `index`.
+    pub fn remove_component_read(&mut self, index: T) {
+        let sparse_set_index = index.sparse_set_index();
+        self.remove_component_sparse_set_index_write(sparse_set_index);
+        self.remove_component_sparse_set_index_read(sparse_set_index);
+    }
+
+    /// Removes write access to the component given by `index`.
+    pub fn remove_component_write(&mut self, index: T) {
```
Given the previous soundness issue, in these docs I would advise against calling these methods on `Access`es that might be prepopulated elsewhere, and instead suggest creating a new one and calling `extend`.

pcwalton replied:
I added an explanation based on set operations, which was the easiest way I could think of.

pcwalton commented on crates/bevy_ecs/src/query/access.rs:
```diff
@@ -736,30 +728,37 @@ impl<T: SparseSetIndex> Access<T> {
+    /// The returned flag specifies whether the list consists of the components
+    /// that the access *can* write (false) or whether the list consists of the
+    /// components that the access *can't* write (true).
+    pub(crate) fn component_writes(&self) -> (impl Iterator<Item = T> + '_, bool) {
```
That's because it's marked as `pub(crate)`, and I figured it was fine for the private ECS code to know about the internals of `Access`.

SkiFire13 replied:
I see, that makes sense.

pcwalton commented:
Review comments addressed.

alice-i-cecile reviewed (approved):
I'm happy with the need for this, the implementation and the docs. @chescock, if you'd like to re-review this I'll merge it and let you experiment with optimizations and improvements in follow-up so we don't block the animation work.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1685 if you'd like to help out.
